spec/ics-2-consensus-verification/README.md

```diff
@@ -0,0 +1,310 @@
+state to verify future roots, stored in one blockchain to verify the state of the other.
+Defined as
+
+```go
+type ConsensusState struct {
+  Base VerifierBase
+  Root AccumulatorRoot
+}
+```
+where
+  * `Base` is a data used by `Consensus.Verifier` to verify `Header`s
+  * `Root` is the `AccumuatorRoot`, used to prove internal state
+
+##### Header
+
+`Header` is a blockchain header which provides information to update `ConsensusState`,
+submitted to one blockchain to update the stored `ConsensusState`.
+Defined as
+
+```go
+type Header struct {
+  Proof HeaderProof
+  Base Maybe[VerifierBase]
+  Root AccumulatorRoot
+}
+```
+where
+  * `Proof` is the commit proof used by `Consensus.Verifier` to be verified
+  * `Base` is the new verify, if needed to be updated
+  * `Root` is the new `AccumulatorRoot` which will replace the existing one
+
+##### Verifier
+
+`Verifier` is a light client verifier proving `Header` depending on the `Commit`.
+It is expected to prove far more efficiently than replaying `Consensus` logic
+for the given parent `Header` and the list of network messages, idealy in O(1) time.
+Defined as
+
+```go
+type Verifier func(ConsensusState, Header) (Error|ConsensusState)
+```
+
+##### Commit
+
+*// Commit and Blockchain is not used inside the protocol, but for the understanding
+of how does verifiers, cstates, headers work & generated. I think we should move these
+to another paragraph, as the other types and functions are actually used by other icss*
+
+`Commit` is a `Header` generating function which takes the previous `Header` with the messages
+and returns the result.
+
+```go
+type Commit func(ConsensusState, [Message]) Header
+```
+
+##### Blockchain
+
+Defined as blockchain consensus algorithm which generates valid `Header`s.
+It generatess a unique list of headers starting from a genesis `ConsensusState` with arbitrary
+messages.
+
+`Blockchain` is defined as
+```go
+type Blockchain struct {
+  Verifier Verifier
+  Genesis ConsensusState
+  Commit Commit
+}
+```
+where
+  * `Verifier` is the lightclient verifier
+  * `Genesis` is the genesis `ConsensusState`
+  * `Commit` is the header generating function
+
+The headers generated from the `Blockchain` is expected to satisfy the followings:
+
+1. The `Header`s have no more than one direct child
+
+* Satisfied if: deterministic safety
+* Possible violation scenario: validator double signing, chain reorganization (Nakamoto consensus)
+
+2. The `Header`s eventually have at least one direct child
+
+* Satisfied if: liveness, light-client verifier continuity
+* Possible violation scenario: synchronised halt, incompatible hard fork
+
+3. The `Header`s are generated from the `Commit`, which ensures valid transition of the state
+
+* Satisfied if: correct block generation & state machine
+* Possible violation scenario: invariant break, validator cartel
+
+In these cases it is ensured that the `Verifier` works as intended.
+
```

Comment:
I still think we want to talk about safety in terms of the light client reflecting what the full nodes do

"A light client validates a header iff. validators came to consensus on it" is a stronger statement, and it encapsulates no more than one direct child (finality), at least one direct child (liveness) - since those are both properties of the consensus (e.g. Tendermint)

Reply:
Applied, also moved to VERIFIER.md